OverlappingFileLockException when urlmap and paths are triggered sequentially

When I request urlmap and immediately after getting result paths for the same root in the same workspace, then paths ends up by an error. It seems that the graph connection was closed (FileNeo4JGraphConnection.close() method was invoked) but the graph is still locked in the filesystem.

We can workaround it by catching the exception, wait for some period of time and then retry. Or is there any better solution to this?

Also tried newer versions of Neo4j, but it wasn't any better.

```
[XNIO-1 task-3] ERROR io.undertow.request - UT005023: Exception handling request to /api/depgraph/graph/paths
org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.StoreLockerLifecycleAdapter@5578e898' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:76) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:212) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:149) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:372) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:179) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:220) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:56) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[jboss-servlet-api_3.1_spec-1.0.0.Final.jar:1.0.0.Final]
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:130) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at org.commonjava.aprox.bind.jaxrs.ResourceManagementFilter.doFilter(ResourceManagementFilter.java:60) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.aprox.bind.jaxrs.ResourceManagementFilter$Proxy$_$$_WeldClientProxy.doFilter(Unknown Source) ~[aprox-embedder-savant-0.25.1.jar:na]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:60) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:132) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:85) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:61) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:56) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:45) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:63) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:58) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:70) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.security.handlers.SecurityInitialHandler.handleRequest(SecurityInitialHandler.java:76) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at org.commonjava.aprox.bind.jaxrs.HeaderDebugger.handleRequest(HeaderDebugger.java:93) ~[aprox-embedder-savant-0.25.1.jar:na]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:261) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:247) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:76) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:166) ~[undertow-servlet-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:197) ~[undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:764) [undertow-core-1.1.2.Final.jar:1.1.2.Final]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_60]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
Caused by: java.lang.RuntimeException: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.StoreLockerLifecycleAdapter@5578e898' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.kernel.InternalAbstractGraphDatabase.run(InternalAbstractGraphDatabase.java:287) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.EmbeddedGraphDatabase.<init>(EmbeddedGraphDatabase.java:106) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.graphdb.factory.GraphDatabaseFactory$1.newDatabase(GraphDatabaseFactory.java:88) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.graphdb.factory.GraphDatabaseBuilder.newGraphDatabase(GraphDatabaseBuilder.java:207) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.graphdb.factory.GraphDatabaseFactory.newEmbeddedDatabase(GraphDatabaseFactory.java:69) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.commonjava.maven.atlas.graph.spi.neo4j.FileNeo4JGraphConnection.<init>(FileNeo4JGraphConnection.java:148) ~[atlas-driver-neo4j-embedded-0.15.2.jar:na]
	at org.commonjava.maven.atlas.graph.spi.neo4j.FileNeo4jConnectionFactory.openConnection(FileNeo4jConnectionFactory.java:93) ~[atlas-driver-neo4j-embedded-0.15.2.jar:na]
	at org.commonjava.maven.atlas.graph.RelationshipGraphFactory.open(RelationshipGraphFactory.java:71) ~[atlas-relationships-api-0.15.2.jar:na]
	at org.commonjava.cartographer.graph.GraphResolver.resolveGraph(GraphResolver.java:313) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.cartographer.graph.GraphResolver.resolveToGraphMap(GraphResolver.java:258) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.cartographer.graph.GraphResolver.resolveAndExtractMultiGraph(GraphResolver.java:152) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.cartographer.INTERNAL.ops.GraphOpsImpl.getPaths(GraphOpsImpl.java:139) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.aprox.depgraph.rest.GraphController.getPaths(GraphController.java:85) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.aprox.depgraph.jaxrs.GraphResource.getPaths(GraphResource.java:56) ~[aprox-embedder-savant-0.25.1.jar:na]
	at org.commonjava.aprox.depgraph.jaxrs.GraphResource$Proxy$_$$_WeldClientProxy.getPaths(Unknown Source) ~[aprox-embedder-savant-0.25.1.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_60]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_60]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_60]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_60]
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:137) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:296) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:250) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:237) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:356) ~[resteasy-jaxrs-3.0.9.Final.jar:na]
	... 34 common frames omitted
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.StoreLockerLifecycleAdapter@5578e898' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:497) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:104) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.InternalAbstractGraphDatabase.run(InternalAbstractGraphDatabase.java:265) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	... 57 common frames omitted
Caused by: org.neo4j.kernel.StoreLockException: Unable to obtain lock on store lock file: /home/petr/temp/rtgov/aprox/var/lib/aprox/data/depgraph/ant-1.8.0/store_lock. Please ensure no other process is using this database, and that the directory is writable (required even for read-only access)
	at org.neo4j.kernel.StoreLocker.checkLock(StoreLocker.java:74) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.StoreLockerLifecycleAdapter.start(StoreLockerLifecycleAdapter.java:40) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:491) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	... 59 common frames omitted
Caused by: java.nio.channels.OverlappingFileLockException: null
	at sun.nio.ch.SharedFileLockTable.checkList(FileLockTable.java:255) ~[na:1.8.0_60]
	at sun.nio.ch.SharedFileLockTable.add(FileLockTable.java:152) ~[na:1.8.0_60]
	at sun.nio.ch.FileChannelImpl.tryLock(FileChannelImpl.java:1107) ~[na:1.8.0_60]
	at java.nio.channels.FileChannel.tryLock(FileChannel.java:1155) ~[na:1.8.0_60]
	at org.neo4j.kernel.impl.nioneo.store.FileLock.wrapFileChannelLock(FileLock.java:35) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.impl.nioneo.store.FileLock.getOsSpecificFileLock(FileLock.java:93) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.DefaultFileSystemAbstraction.tryLock(DefaultFileSystemAbstraction.java:83) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	at org.neo4j.kernel.StoreLocker.checkLock(StoreLocker.java:70) ~[neo4j-kernel-1.9.6.jar:1.9.6]
	... 61 common frames omitted
```

--- a/drivers/neo4j-embedded/src/main/java/org/commonjava/maven/atlas/graph/spi/neo4j/FileNeo4jConnectionFactory.java
+++ b/drivers/neo4j-embedded/src/main/java/org/commonjava/maven/atlas/graph/spi/neo4j/FileNeo4jConnectionFactory.java
@@ -20,11 +20,14 @@ import org.apache.commons.lang.StringUtils;
 import org.commonjava.maven.atlas.graph.spi.RelationshipGraphConnection;
 import org.commonjava.maven.atlas.graph.spi.RelationshipGraphConnectionException;
 import org.commonjava.maven.atlas.graph.spi.RelationshipGraphConnectionFactory;
+import org.neo4j.kernel.StoreLockException;
+import org.neo4j.kernel.lifecycle.LifecycleException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.channels.OverlappingFileLockException;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
@@ -41,6 +44,8 @@ public class FileNeo4jConnectionFactory
     private final File dbBaseDirectory;
 
     private final boolean useShutdownHook;
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
 
     private int storageBatchSize = FileNeo4JGraphConnection.DEFAULT_BATCH_SIZE;
 
@@ -90,7 +95,40 @@ public class FileNeo4jConnectionFactory
         FileNeo4JGraphConnection conn = openConnections.get( workspaceId );
         if ( conn == null || !conn.isOpen() )
         {
-            conn = new FileNeo4JGraphConnection( workspaceId, db, useShutdownHook, storageBatchSize, this );
+            conn = null;
+            int attempt = 0;
+
+            while ( conn == null )
+            {
+                attempt++;
+                try
+                {
+                    conn = new FileNeo4JGraphConnection( workspaceId, db, useShutdownHook, storageBatchSize, this );
+                }
+                catch ( RuntimeException ex )
+                {
+                    if ( ex.getCause() instanceof LifecycleException
+                            && ex.getCause().getCause() instanceof StoreLockException
+                            && ex.getCause().getCause().getCause() instanceof OverlappingFileLockException
+                            && attempt < 3 )
+                    {
+                        logger.warn( "Tried to connect to DB which is not closed (yet). {} Retrying in 5s.", ex.toString() );
+                        try
+                        {
+                            Thread.sleep(5000);
+                        }
+                        catch ( InterruptedException ez )
+                        {
+                            logger.error( "The wait delay was interrupted.", ex );
+                        }
+                    }
+                    else
+                    {
+                        throw ex;
+                    }
+                }
+            }
+
             openConnections.put( workspaceId, conn );
         }
 

